### ISSUE TITLE
If a job fails because it "has been attempted too many times", the original error never gets reported

### ISSUE THREAD

**Issue by tkoop** [laravel/horizon#1514]
### Horizon Version

5.29.0

### Laravel Version

11.10.0

### PHP Version

8.2

### Redis Driver

PhpRedis

### Redis Version

7.2.4

### Description

In our logs, we are noticing that a job is failing. It says this job "has been attempted too many times". But the stack trace in the log only shows the stack trace for `/vendor/laravel/framework/src/Illuminate/Queue/MaxAttemptsExceededException.php:24`, not the original error. I want to know the real reason the job is failing, and the stack trace and exception of the original error.

### Steps To Reproduce

This is our horizon.php config file.
```
<?php

use Illuminate\Support\Str;

return [
	/*
    |--------------------------------------------------------------------------
    | Horizon Domain
    |--------------------------------------------------------------------------
    |
    | This is the subdomain where Horizon will be accessible from. If this
    | setting is null, Horizon will reside under the same domain as the
    | application. Otherwise, this value will serve as the subdomain.
    |
    */

	"domain" => env("HORIZON_DOMAIN"),

	/*
    |--------------------------------------------------------------------------
    | Horizon Path
    |--------------------------------------------------------------------------
    |
    | This is the URI path where Horizon will be accessible from. Feel free
    | to change this path to anything you like. Note that the URI will not
    | affect the paths of its internal API that aren't exposed to users.
    |
    */

	"path" => env("HORIZON_PATH", "horizon"),

	/*
    |--------------------------------------------------------------------------
    | Horizon Redis Connection
    |--------------------------------------------------------------------------
    |
    | This is the name of the Redis connection where Horizon will store the
    | meta information required for it to function. It includes the list
    | of supervisors, failed jobs, job metrics, and other information.
    |
    */

	"use" => "default",

	/*
    |--------------------------------------------------------------------------
    | Horizon Redis Prefix
    |--------------------------------------------------------------------------
    |
    | This prefix will be used when storing all Horizon data in Redis. You
    | may modify the prefix when you are running multiple installations
    | of Horizon on the same server so that they don't have problems.
    |
    */

	"prefix" => env("HORIZON_PREFIX", Str::slug(env("APP_NAME", "laravel"), "_") . "_horizon:"),

	/*
    |--------------------------------------------------------------------------
    | Horizon Route Middleware
    |--------------------------------------------------------------------------
    |
    | These middleware will get attached onto each Horizon route, giving you
    | the chance to add your own middleware to this list or change any of
    | the existing middleware. Or, you can simply stick with this list.
    |
    */

	"middleware" => ["web"],

	/*
    |--------------------------------------------------------------------------
    | Queue Wait Time Thresholds
    |--------------------------------------------------------------------------
    |
    | This option allows you to configure when the LongWaitDetected event
    | will be fired. Every connection / queue combination may have its
    | own, unique threshold (in seconds) before this event is fired.
    |
    */

	"waits" => [
		"redis:default" => 60,
	],

	/*
    |--------------------------------------------------------------------------
    | Job Trimming Times
    |--------------------------------------------------------------------------
    |
    | Here you can configure for how long (in minutes) you desire Horizon to
    | persist the recent and failed jobs. Typically, recent jobs are kept
    | for one hour while all failed jobs are stored for an entire week.
    |
    */

	"trim" => [
		"recent" => 60,
		"pending" => 60,
		"completed" => 60,
		"recent_failed" => 10080,
		"failed" => 10080,
		"monitored" => 10080,
	],

	/*
    |--------------------------------------------------------------------------
    | Silenced Jobs
    |--------------------------------------------------------------------------
    |
    | Silencing a job will instruct Horizon to not place the job in the list
    | of completed jobs within the Horizon dashboard. This setting may be
    | used to fully remove any noisy jobs from the completed jobs list.
    |
    */

	"silenced" => [
		// App\Jobs\ExampleJob::class,
	],

	/*
    |--------------------------------------------------------------------------
    | Metrics
    |--------------------------------------------------------------------------
    |
    | Here you can configure how many snapshots should be kept to display in
    | the metrics graph. This will get used in combination with Horizon's
    | `horizon:snapshot` schedule to define how long to retain metrics.
    |
    */

	"metrics" => [
		"trim_snapshots" => [
			"job" => 24,
			"queue" => 24,
		],
	],

	/*
    |--------------------------------------------------------------------------
    | Fast Termination
    |--------------------------------------------------------------------------
    |
    | When this option is enabled, Horizon's "terminate" command will not
    | wait on all of the workers to terminate unless the --wait option
    | is provided. Fast termination can shorten deployment delay by
    | allowing a new instance of Horizon to start while the last
    | instance will continue to terminate each of its workers.
    |
    */

	"fast_termination" => false,

	/*
    |--------------------------------------------------------------------------
    | Memory Limit (MB)
    |--------------------------------------------------------------------------
    |
    | This value describes the maximum amount of memory the Horizon master
    | supervisor may consume before it is terminated and restarted. For
    | configuring these limits on your workers, see the next section.
    |
    */

	"memory_limit" => 64,

	/*
    |--------------------------------------------------------------------------
    | Queue Worker Configuration
    |--------------------------------------------------------------------------
    |
    | Here you may define the queue worker settings used by your application
    | in all environments. These supervisors and settings handle all your
    | queued jobs and will be provisioned by Horizon during deployment.
    |
    */

	"defaults" => [],

	"environments" => [
		// in dev, just have one queue worker for everything
		"local" => [
			"supervisor-default" => [
				"connection" => "redis",
				"queue" => [
					"default",
					"notifications",
					"notifications-priority",
					"notifications-bulk",
					"background-processes",
				],
				"balance" => false,
				"minProcesses" => 1,
				"maxProcesses" => 1,
				"timeout" => config("app.browsershot_timeout"),
				"memory" => 128,
				"tries" => 1,
			],
		],

		"*" => [
			// supervisor that manages the default queue and other medium priority tasks
			"supervisor-default" => [
				"connection" => "redis",
				"queue" => ["default"],
				"balance" => false,
				"minProcesses" => 1,
				"maxProcesses" => 1,
				"tries" => 3,
			],

			// supervisor that manages queues that are high priority
			"supervisor-notifications" => [
				"connection" => "redis",
				"queue" => ["notifications", "notifications-priority", "notifications-bulk"],
				"balance" => "simple",
				"minProcesses" => 1,
				"maxProcesses" => 3,
				"tries" => 3,
			],

			// supervisor that manages processes which can use a lot of resources
			"supervisor-background" => [
				"connection" => "redis",
				"queue" => ["background-processes"],
				"balance" => false,
				"minProcesses" => 1,
				"maxProcesses" => 1,
				"timeout" => config("app.browsershot_timeout"),
				"memory" => 128,
				"tries" => 3,
			],
		],
	],
];
```

**Comment by crynobone** [comment]
Feel free to send a PR if you want this feature.

**Comment by javis** [comment]
Having the same issue since upgrading to Laravel 11. I suspect Horizon is having a bug at killing gracefully the process that has to be terminated

**Comment by marensas999** [comment]
It's a bug not a feature.

**Comment by benjibee** [comment]
> It's a bug not a feature.

I'd like to second this, it's quite a big issue. The jobs I've had fail where the only visible exception is that the job was tried to many times are painful to debug and track down. The reason I'd say this is a bug and not a feature is that reporting that the job failed because it was tried too many times is simply not reporting the reason the job failed. The exception is visible and right there, but not reported.

**Comment by crynobone** [comment]
Feel free to submit a PR to fix the issue. Bug or feature this is an open source project where the community can contribute.

**Comment by benjibee** [comment]
> Feel free to submit a PR to fix the issue. Bug or feature this is an open source project where the community can contribute.

I looked into it before posting my comment. It's far outside of my abilities at the moment. Too many areas I've wholly unfamiliar with. If it's not a priority for the team then maybe I'll get there at some point and then I'd be happy to contribute.

**Comment by JavoByte** [comment]
I don't think this issue should have been closed.

It would be nice to have it open to give visibility of known issues.

I also have this issue and I have no clue why my jobs could be failing. 

> Feel free to submit a PR to fix the issue. Bug or feature this is an open source project where the community can contribute.

while this is true, not every developer has the necessary expertise or familiarity with the codebase to just submit a PR. What would be the purpose of reporting issues then?

**Comment by luke-hill** [comment]
I get that this issue may not be high priority e.t.c. for laravel developers. But is it "completed". i.e. fixed?

By all means it may not get looked at, but I'd argue it should remain open

**Comment by MitchTalmadge** [comment]
@crynobone Please re-open this issue, it was not completed.

**Comment by ugurarici** [comment]
I’m puzzled as to why my job keeps being retried. When I call `dispatchSync()` in a command, it executes without issue, but when I use `dispatch()` to push it onto the queue, it fails with:

`Illuminate\Queue\MaxAttemptsExceededException: App\Jobs\XXX has been attempted too many times.`

What causes it to retry in the first place? I need a guide at least :)

**Comment by artem-schander** [comment]
I have wasted three days to fix one of my jobs that behave like this. I had to log and debug a lot. It was very frustrating and I could exclude many possible causes. It was not memory nor timeout related. I have still no idea what really caused it. Its (hopefully) gone since i reduced the number of (async) event dispatches and the excessive use of cache. in the end it made no sence that this changes had this kind of impact on the ability to finish this jobs.
unfortunately i have some others that are also failing with the "has been attempted too many times" message and i really dont feel like fixing them :(

Ps.
I would expect to see this "has been attempted too many times" message only when there was no catchable exception. For example when oom just killst the process, but this is definitely not the case here

**Comment by marensas999** [comment]
Fixed by entering Redis CLI `redis-cli` and searching for `keys *overlap*` and deleted them `DEL <key>`.

It helps temporarily for some time and then it happens again.

It's related with "WithoutOverlapping" middleware I guess.

**Comment by JohnnyWalkerDigital** [comment]
Community: "I've found an issue with your software"
Project devs: "Feel free to fix it - bye!" ISSUE CLOSED

**Comment by csalmeida** [comment]
We tried to fix this issue for days for our product Contactzilla and in our case it was the `queue.php` config.
The `retry_after` settings needs to be greater than the duration of the batched job.

```yml
    'redis' => [
        'driver' => 'redis',
        'connection' => 'default',
        'queue' => env('REDIS_QUEUE', 'default'),
        'retry_after' => 9000, // <-- Set this to a higher value (default was 90)
        'block_for' => null,
        'after_commit' => false,
    ],
```

**Comment by drakakisgeo** [comment]
Have the same problem, I also use the middleware withoutOverlaping to the job tha failing like this. When I disabled everything worked fine.

**Comment by lasselehtinen** [comment]
I've bumped to the same issue soon as I added WithoutOverlapping middleware to bunch of jobs. I've checked the that the timeout configuration should be correct, retry_after > timeout. I have no job spesific timeouts defined. 

```
> config('queue.connections.redis.retry_after');
= 7800

> config('horizon.waits.redis:default');
= 7500

> config('horizon.environments.worker.supervisor-1.timeout');
= 7200
```

The strangest that when you queue:retry these jobs that have thrown MaxAttemptsExceededException, they fail again instantly with the same exception. I think normally it resets the attemps count back to 0 when job is placed in the failed jobs. 

```
2025-10-01 10:38:07 App\Jobs\UpdateIndexJob ..................... RUNNING
2025-10-01 10:38:07 App\Jobs\UpdateIndexJob ................. 1.77ms FAIL
2025-10-01 10:38:07 App\Jobs\UpdateIndexJob ..................... RUNNING
2025-10-01 10:38:07 App\Jobs\UpdateIndexJob ................. 1.70ms FAIL
2025-10-01 10:38:07 App\Jobs\UpdateIndexJob ..................... RUNNING
2025-10-01 10:38:07 App\Jobs\UpdateIndexJob ................. 1.65ms FAIL
2025-10-01 10:38:07 App\Jobs\UpdateIndexJob ..................... RUNNING
2025-10-01 10:38:07 App\Jobs\UpdateIndexJob ................. 1.62ms FAIL
2025-10-01 10:38:07 App\Jobs\UpdateIndexJob ..................... RUNNING
2025-10-01 10:38:07 App\Jobs\UpdateIndexJob ................. 1.58ms FAIL
```
I also don't have the skills to deep dive into this, but it has definately something to do with the WithoutOverlapping middleware soon as we added it. I was just thinking when this overlapping happens and it is release back the queue, it counts it as an attempt? And therefore we might be hitting the tries limit quite quickly. And then this is combined with the fact that it seems that the attempts count is not resetted after it is placed in the failed jobs.

**Comment by fabioselau077** [comment]
> I've bumped to the same issue soon as I added WithoutOverlapping middleware to bunch of jobs. I've checked the that the timeout configuration should be correct, retry_after > timeout. I have no job spesific timeouts defined.
> 
> ```
> > config('queue.connections.redis.retry_after');
> = 7800
> 
> > config('horizon.waits.redis:default');
> = 7500
> 
> > config('horizon.environments.worker.supervisor-1.timeout');
> = 7200
> ```
> 
> The strangest that when you queue:retry these jobs that have thrown MaxAttemptsExceededException, they fail again instantly with the same exception. I think normally it resets the attemps count back to 0 when job is placed in the failed jobs.
> 
> ```
> 2025-10-01 10:38:07 App\Jobs\UpdateIndexJob ..................... RUNNING
> 2025-10-01 10:38:07 App\Jobs\UpdateIndexJob ................. 1.77ms FAIL
> 2025-10-01 10:38:07 App\Jobs\UpdateIndexJob ..................... RUNNING
> 2025-10-01 10:38:07 App\Jobs\UpdateIndexJob ................. 1.70ms FAIL
> 2025-10-01 10:38:07 App\Jobs\UpdateIndexJob ..................... RUNNING
> 2025-10-01 10:38:07 App\Jobs\UpdateIndexJob ................. 1.65ms FAIL
> 2025-10-01 10:38:07 App\Jobs\UpdateIndexJob ..................... RUNNING
> 2025-10-01 10:38:07 App\Jobs\UpdateIndexJob ................. 1.62ms FAIL
> 2025-10-01 10:38:07 App\Jobs\UpdateIndexJob ..................... RUNNING
> 2025-10-01 10:38:07 App\Jobs\UpdateIndexJob ................. 1.58ms FAIL
> ```
> 
> I also don't have the skills to deep dive into this, but it has definately something to do with the WithoutOverlapping middleware soon as we added it. I was just thinking when this overlapping happens and it is release back the queue, it counts it as an attempt? And therefore we might be hitting the tries limit quite quickly. And then this is combined with the fact that it seems that the attempts count is not resetted after it is placed in the failed jobs.

I have the same problem using WithoutOverlapping. I even thought it was a problem with retry, but I debugged all the configurations and the job itself and nothing was wrong.

**Comment by mkamra2000** [comment]
Yes, I faced the same issue. The job **attempt count kept increasing**, but the **original exception was not logged anywhere**, which made debugging very difficult.

<img width="1919" height="961" alt="Image" src="https://github.com/user-attachments/assets/0c5a85ec-931c-4f50-98dd-988a0543bc4d" />

In my case, the issue was resolved after changing the Redis serializer from
`Redis::SERIALIZER_PHP` to `Redis::SERIALIZER_NONE`.

### Fix

```php
'queue' => [
    'url' => env('REDIS_URL'),
    'host' => env('REDIS_HOST', '127.0.0.1'),
    'username' => env('REDIS_USERNAME', 'default'),
    'password' => env('REDIS_PASSWORD'),
    'port' => env('REDIS_PORT', '1234'),
    'database' => env('REDIS_QUEUE_DB', '0'),
    'scheme' => env('REDIS_SCHEME', 'tls'),
    'options' => [
        'cluster' => false,
        'prefix' => env(
            'REDIS_QUEUE_PREFIX',
            Str::slug(env('APP_NAME', 'laravel'), '_') . '_' . env('APP_ENV', 'local') . '_queue_'
        ),
        'serializer' => class_exists('Redis')
            ? \Redis::SERIALIZER_NONE
            : 0,
    ],
],
```

### Why this works

When `Redis::SERIALIZER_PHP` is used, certain job payloads can fail during unserialization inside the worker.
This results in:

* Jobs being retried continuously
* `attempts` increasing
* No meaningful exception appearing in logs or Horizon UI

Switching to `Redis::SERIALIZER_NONE` avoids PHP-level serialization and lets Laravel handle payload encoding consistently, preventing these silent failures.

### References

* Laravel framework issue:
  [https://github.com/laravel/framework/issues/47356](https://github.com/laravel/framework/issues/47356)
* Related fix / discussion:
  [https://github.com/laravel/framework/pull/48180](https://github.com/laravel/framework/pull/48180)
